src/main/java/io/openliberty/tools/common/plugins/util/DevUtil.java

```diff
@@ -566,39 +644,366 @@ public void onFileChange(File file) {
             }
             long serverStartTimeoutMillis = serverStartTimeout * 1000;
-
             // Wait for the server started message in messages.log
             String startMessage = serverTask.waitForStringInLog(START_SERVER_MESSAGE_PREFIX, serverStartTimeoutMillis,
                     messagesLogFile);
             if (startMessage == null) {
                 setDevStop(true);
-                stopServer();
+                if (container) {
+                    stopContainer();
+                } else {
+                    stopServer();
+                }
                 throw new PluginExecutionException("The server has not started within " + serverStartTimeout + " seconds. " +
                         "Consider increasing the server start timeout if this continues to occur. " +
                         "For example, " + getServerStartTimeoutExample());
             }
-
             // Check for port already in use error
             String portError = serverTask.findStringInFile(PORT_IN_USE_MESSAGE_PREFIX, messagesLogFile);
             if (portError != null) {
                 error(portError.split(PORT_IN_USE_MESSAGE_PREFIX)[1]);
             }
-
             // Parse hostname, http, https ports for integration tests to use
             parseHostNameAndPorts(serverTask, messagesLogFile);
         } catch (IOException e) {
             throw new PluginExecutionException("An error occurred while starting the server: " + e.getMessage(), e);
         }
     }
 
+    private List<String> readDockerfile(File dockerfile) throws PluginExecutionException {
+        // Convert Dockerfile to List of strings for each line
+        List<String> dockerfileLines = null;
+        try {
+            dockerfileLines = Files.readAllLines(dockerfile.toPath());
+        } catch (IOException e) {
+            error("Failed to read Dockerfile located at " + dockerfile);
+            throw new PluginExecutionException("Could not read Dockerfile " + dockerfile + ": " + e.getMessage(), e);
+        }
+        return dockerfileLines;
+    }
+
+    private List<String> removeWarFileLines(List<String> dockerfileLines) {
+        List<String> warFileLines = new ArrayList<String>();
+        for (String line : dockerfileLines) {
+            // Remove white space from the beginning and end of the line
+            String trimLine = line.trim();
+            if (!trimLine.startsWith("#") && trimLine.toLowerCase().contains(".war")) {
+                // Break the Dockerfile line down into segments based on any amount of whitespace.
+                // The command must be to the left of any comments.
+                String[] cmdSegments = trimLine.split("#")[0].split("\\s+");
+                // if the line starts with COPY and the second to last segment ends with ".war", it is a WAR file COPY line
+                if (cmdSegments[0].equalsIgnoreCase("COPY") && cmdSegments[cmdSegments.length - 2].toLowerCase().endsWith(".war")) {
```

Comment:
What if the line does not have 3 segments? The `cmdSegments[cmdSegments.length-2]` part of this will get an ArrayIndexOutOfBoundsException.

Reply:
This is checked in the part where it has:
```
                    if (cmdSegments.length < 3) {
                        throw new PluginExecutionException("Incorrect syntax on this line in the Dockerfile: '" + line + 
                        "'. There must be at least two arguments for the COPY command, a source path and a destination path.");
                    }
```

Reply:
That check is in another method and does not apply to here. This line will fail with the exception before it ever gets to that other method.

Reply:
Added validation to this method